---
title: 异常和错误处理
date: 2020-05-10 01:42:26
tags:
---
## 异常和错误处理

异常处理 是一种可控的方式处理错误的机制。 **并非处理预期的错误，而是预期之外的错误。**  
例子：无效的邮件地址（预期的错误）  磁盘空间不足，服务不可用（预期之外的错误）
### 1 Error对象和try…catch
可以用来处理任意类型的错误（异常或预期错误）。  
执行代码期间可能会发生的错误有多种类型。每种错误都有对应的错误类型，而当错误发生时，就 会抛出相应类型的错误对象。js共定义了下列 7 种错误类型：
* Error ‰
* EvalError
* RangeError ‰
* ReferenceError ‰
* SyntaxError ‰
* TypeError ‰
* URIError  
Error类型是基本的错误类型，其他类型都继承自这个类型。
```js
try { 
  someFunction(); 
} catch (error){ 
  if (error instanceof TypeError){ 
    //处理ૌ类型错误 
  } else if (error instanceof ReferenceError){ 
    //处理引用错误 
  } else { 
      //处理其他的错误 
  }
}
```
把有可能出的问题的代码放在 try 语句中。try语句中可以理论上可以写任何的代码，只要有一行代码出现问题，整个程序的执行流程就会立即调到catch语句中执行。  
一旦try中有一行代码发生异常，则这行出错代码的后面的try中的其他语句都不会再执行。   

在执行catch中的代码之前，js引擎会首先根据错误类型自动创建一个错误，并通过catch后面的参数传递到catch中。不同的浏览器创建的error对象不一样，但是同创他们都包含一个message属性，值是这个错误的一些信息。  
catch中的代码执行完毕之后，会继续执行后面的代码，程序不会停止下来。
##### 合理使用try……catch
​ 当 try-catch 语句中发生错误时，浏览器会认为错误已经被处理了，浏览器就不再报告错误了。这也是最简单的一种情况。

​ 使用 try-catch 最适合处理那些我们无法控制的错误。假设你在使用一个大型 JavaScript 库中的 函数，该函数可能会有意无意地抛出一些错误。由于我们不能修改这个库的源代码，所以大可将对该函 数的调用放在 try-catch 语句当中，一有什么错误发生，也好可以恰当地处理它们。

​ 在明明知道自己的代码会发生错误时，再使用 try-catch 语句就不太合适了。例如，如果 传给函数的参数是字符串而非数值，就会造成函数出错，那么就应该先检查参数的类型，然后再决定 如何去做。在这种情况下，不应用使用 try-catch 语句。因为try…catch语句比较是比较好资源的事情
#### finally  
很多时候，try块中包含一些对资源的引用，不管有没有发生错误，都需要释放这些资源，防止应用程序永远占用着资源，由于try块中可与包含很多语句，任何一条都有可能发生错误，所以在try块里释放资源并不安全（因为错误可能发生在释放资源之前，这样就没有机会释放资源了）。同样在catch中释放资源也不安全，因为catch中的代码只有在发生错误的时候才会执行。这时finally就派上用场了，不管是否发生错误，finally中的代码都会被执行
```js

try{
    console.log("this line is executed....");
    throw new Error("whoops");
    console.log("this line is not...");
}catch(err){
    console.log("there was an error...");
}finally {
     console.log("...always executed");
     console.log("perform cleanup here");
}
```
### 主动抛出异常
在大部分的代码执行过程中，都是出现错误的时候，由浏览器(javascript引擎)抛出异常，然后程序或者停止执行，或被try…catch 捕获。

​ 然而有时候我们在检测到一些不合理的情况发生的时候也可以主动抛出错误。

​ 使用 throw 关键字抛出来主动抛出异常。
```js
    function foo(num) {
        if(typeof num == "number"){
            return num * num;
        }else{
            throw new TypeError("类型错误，你应该传入一个数字...")
        }
    }
    console.log(foo(33))
    console.log(foo("abc"))
```
### JavaScript 异常的收集
异常出现以后，会触发错误事件传递给window对象， 可以用 window.onerror = function(){}来捕抓页面上每一个错误，所以捕抓错误很简单。  
我们要把我们需要的错误信息，post去我们的日志记录服务器那里，要做的：

整理需要记录的信息（包括，error属性，userAgent等）  
error兼容性处理，不同浏览器error对象的属性会有差异，要做兼容处理  
属性丢失，我们会想要收集些比较有用的信息包括：错误类型（name）、错误消息（message）、脚本文件地址（script）、行号（line)、列号（column）、堆栈跟踪（stack）。如果一个异常是通过 try-catch 捕获到的，这些信息都在 Error 对象上（主流浏览器都支持）。但如果是通过 window.onerror 捕获到的，我们都知道这个事件函数只有 3 个参数，所以这 3 个参数以外的信息就丢失了。  
更多相关：https://www.jianshu.com/p/315ffe6797b8
